core/src/verified_vote_packets.rs

```diff
@@ -128,7 +151,28 @@ impl<'a> Iterator for ValidatorGossipVotesIterator<'a> {
     }
 }
 
-pub type SingleValidatorVotes = BTreeMap<(Slot, Hash), (PacketBatch, Signature)>;
+pub enum SingleValidatorVotes {
+    FullVote((Slot, Hash), (PacketBatch, Signature)),
```

Comment:
I think the problem with only keeping the latest `FullVote` is that potentially the latest vote is for the wrong fork that will then get filtered out here: https://github.com/solana-labs/solana/blob/cc767411b24c9d69cadc0f498dcc09d5babfa8b0/core/src/verified_vote_packets.rs#L85-L89.

For now, might be safer to keep the votes around and then send the latest one that is on the right fork for that particular leader's bank

Reply:
So the issue is it depends on which fork this validator makes his bank on right?

At the time he receives the vote, he doesn't know which fork his bank will be on, so we probably don't want to discard some votes that:
1. Aren't the highest slot globally
2. But would be the highest slot on some other fork that the validator then makes a bank on

Reply:
Since it's not the highest slot globally that must mean the validator has then voted on a different fork. If that's the case  do we need to land the validators old vote in the old fork?

Reply:
hmm this is a good question. I'm not sure, I'll have to think it over some more. Seems like it might be ok at first glance, but might be missing an edge case.

Reply:
Thinking about switching proofs. Is it ever essential for your vote for a smaller slot to land on that fork in order to unblock a switching proof for somebody else?

Reply:
Think it's ok, b/c in order for you to have picked a higher fork to vote on that fork must be heavier, so nobody should be blocked trying to switch elsewhere...

Reply:
Yeah that's what I was thinking. The validators old vote wouldn't enable any switching that isn't currently enabled by the latest global vote, if we need to show greater than 1/3 locked out, the new global vote should be good enough to do that. 

This is also only for sending to bank, the votes will still be counted towards optimistic confirmation as that is handled by a different thread. 

For threshold check since we roll up the stakes to ancestors this wouldn't be an issue either right?